### PR TITLE
feat(ext/node): embed std/node into the snapshot

### PR DESCRIPTION
**Pull request by bartlomieju**
This commit moves "deno_std/node" in "ext/node" crate. The code is
transpiled and snapshotted during the build process.

During the first pass a minimal amount of work was done to create the snapshot,
a lot of code in "ext/node" depends on presence of "Deno" global. This code
will be gradually fixed in the follow up PRs to migrate it to import relevant
APIs from "internal:" modules.

Currently the code from snapshot is not used in any way, and all Node/npm
compatibility still uses code from "https://deno.land/std/node" (or from the
location specified by "DENO_NODE_COMPAT_URL"). This will also be
handled in a follow up PRs.

### CODE REVIEW

**Comment by bartlomieju** [comment]
There are currently 3 blockers that we need to tackle:

1. [ ] `tools/lint.js` passes in `ext/node`
2. [ ] references to `Deno` namespace are replaced with imports of internal modules
3. [ ] replace imports of functionality outside of `std/node` with some embedded code:
 - [ ] `ext/node/polyfills/_test_util.ts` depends on `std/testing/asserts.ts` (@kt3k)
 - [ ] `ext/node/polyfills/assert.ts` depends on `std/testing/asserts.ts` (@kt3k)
 - [ ] `ext/node/polyfills/http.ts` depends on `std/http/http_status.ts`
 - [ ] `ext/node/polyfills/os.ts` depends on `std/fs/eol.ts`
 - [ ] `ext/node/polyfills/url.ts` depends on `std/path/_constants.ts`
 - [ ] `ext/node/polyfills/_crypto/crypto_browserify/asn1.js/test/util.js` depends on `std/testing/asserts.ts`
 - [ ] `ext/node/polyfills/_fs/_fs_open.ts` depends on `std/fs/exists.ts`
 - [ ] `ext/node/polyfills/_fs/_fs_writeFile.ts` depends on `std/streams/write_all.ts`
 - [ ] `ext/node/polyfills/_fs/_fs_writeFile.ts` depends on `std/types.d.ts`
 - [ ] `ext/node/polyfills/internal/buffer.mjs` depends on `std/encoding/base64.ts` (@crowlKats)
 - [ ] `ext/node/polyfills/internal/buffer.mjs` depends on `std/encoding/base64url.ts` (@crowlKats)
 - [ ] `ext/node/polyfills/internal/crypto/hash.ts` depends on `std/crypto/_wasm/mod.ts` (@littledivy)
 - [ ] `ext/node/polyfills/internal/crypto/hash.ts` depends on `std/encoding/hex.ts` (@littledivy)
 - [ ] `ext/node/polyfills/internal/crypto/hash.ts` depends on `std/encoding/base64.ts` (@crowlKats)
 - [ ] `ext/node/polyfills/internal/crypto/hash.ts` depends on `std/encoding/base64url.ts` (@crowlKats)
 - [ ] `ext/node/polyfills/internal_binding/_utils.ts` depends on `std/encoding/base64.ts` (@crowlKats)
 - [ ] `ext/node/polyfills/internal_binding/_utils.ts` depends on `std/encoding/base64url.ts` (@crowlKats)
 - [ ] `ext/node/polyfills/internal_binding/connection_wrap.ts` depends on `std/types.d.ts`
 - [ ] `ext/node/polyfills/internal_binding/node_options.ts` depends on `std/flags/mod.ts`
 - [ ] `ext/node/polyfills/internal_binding/stream_wrap.ts` depends on `std/streams/write_all.ts`
  - [ ] `ext/node/polyfills/internal_binding/stream_wrap.ts` depends on `std/types.d.ts`

**Comment by kt3k** [comment]
I think `ext/node/polyfills/_crypto/crypto_browserify/asn1.js/test/util.js` is testing utility and is not needed to included in ext/node

**Comment by kt3k** [comment]
reference errors or other obvious js errors are resolved. Now the build fails at the assertion `assert_eq!(status, v8::ModuleStatus::Instantiated);` line 1774 of core/runtime.rs. Does this still mean something is wrong in js evaluation?

**Comment by bartlomieju** [comment]
> reference errors or other obvious js errors are resolved. Now the build fails at the assertion `assert_eq!(status, v8::ModuleStatus::Instantiated);` line 1774 of core/runtime.rs. Does this still mean something is wrong in js evaluation?

Yep, something is still going wrong, though it's surprising that we're seeing this. I will debug it and get back to you.

File: core/runtime.rs [hunk 1]
```diff
@@ -1770,6 +1788,14 @@ impl JsRuntime {
...
       .expect("ModuleInfo not found");
     let mut status = module.get_status();
+
+    // FIXME(bartlomieju): this is a hack, when snapshotting we shouldn't
+    // be trying to evaluate same module multiple times
+    if status == v8::ModuleStatus::Evaluated {
+      let (sender, receiver) = oneshot::channel();
+      sender.send(Ok(())).unwrap();
+      return receiver;
+    }
```
Comment (bartlomieju) [inline core/runtime.rs:1798]: This should be fixed in a separate PR, https://github.com/denoland/deno/issues/17764 should resolve it

File: core/runtime.rs [hunk 2]
```diff
@@ -836,15 +836,33 @@ impl JsRuntime {
...
+              receiver_result = receiver => {
+                let result = receiver_result?;
+                // eprintln!("result in receiver {:#?}", result);
+                result?;
+              }
+            }
+            Ok::<(), anyhow::Error>(())
+            // let receiver = self.mod_evaluate(id);
+            // self.run_event_loop(false).await?;
+            // receiver.await?
```
Comment (bartlomieju) [inline core/runtime.rs:865]: These should be reverted

File: core/modules.rs [hunk 3]
```diff
@@ -275,21 +275,27 @@ pub struct NoopModuleLoader;
...
-    _specifier: &str,
-    _referrer: &str,
+    specifier: &str,
+    referrer: &str,
     _kind: ResolutionKind,
   ) -> Result<ModuleSpecifier, Error> {
-    Err(generic_error("Module loading is not supported"))
+    Err(generic_error(
+      format!("Module loading is not supported; attempted to resolve: \"{specifier}\" from \"{referrer}\"")
+    ))
```
Comment (bartlomieju) [inline core/modules.rs:284]: Land in a separate PR

**Review by ry — APPROVED** [review]
LGTM - after landing this we are leaving main in purgatory with all this code being unused. However I think it makes sense given the size of this patch.

**Pull request merged**